### PR TITLE
Fix delete comment doesn't work

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

The delete comment doesn't really delete the comment, so when we reopen the chat, we will see the "deleted" message again.

### Fixed Issues

$ https://github.com/Expensify/App/issues/39432
PROPOSAL: https://github.com/Expensify/App/issues/39432#issuecomment-2033695929

### Tests

Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps
### QA Steps

1. Open any chat and send a message
2. Delete the message that you sent
3. Verify the message is deleted
4. Go to another chat/close the current chat
5. Reopen the chat
6. Verify the deleted message doesn't show back
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/50919443/a8135de3-f802-4bf9-89e7-6cff2f1df991

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/50919443/576940c8-9d1e-4ad7-84bb-9339e78f8ce4

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/50919443/455cbc74-b0e5-4648-9118-e2c862d1283e

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/50919443/6cc28e58-0d04-4513-ace7-a5cb86263411

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/50919443/bcb93db3-6f1a-4817-b433-293df60eb068

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/50919443/071b1065-0d3e-4c50-9274-1c8b1acb7867

</details>

### CODE REVIEW

**Comment by bernhardoj** [comment]
Additional video showing that the add, update, and delete comment isn't processed when doing them while offline.

https://github.com/Expensify/App/assets/50919443/435bd068-e59f-4964-9d7e-9949139dab5e

**Comment by bernhardoj** [comment]
Now, here is the last issue that I found. (online case) If I add a new message and delete it while the add message request is still processing, both add and delete messages are removed from PersistedRequest, but the add message request is still going, so the "deleted" message will show again.

The solution I'm thinking of is if the conflicting request is the ongoing request, we cancel the request too.
```
conflictingRequests.forEach((conflictingRequest) => {
    if (conflictingRequest === processingRequest) {
        HttpUtils.cancelPendingRequests();
    }
```

Here is the diff

<details>
<summary>diff</summary>

```diff
diff --git a/src/libs/Network/SequentialQueue.ts b/src/libs/Network/SequentialQueue.ts
index 38b0549b28..4c5eb62f3c 100644
--- a/src/libs/Network/SequentialQueue.ts
+++ b/src/libs/Network/SequentialQueue.ts
@@ -20,6 +20,7 @@ resolveIsReadyPromise?.();
 let isSequentialQueueRunning = false;
 let currentRequest: Promise<void> | null = null;
 let isQueuePaused = false;
+let processingRequest: OnyxRequest | null = null;
 
 /**
  * Puts the queue into a paused state so that no requests will be processed
@@ -65,6 +66,7 @@ function process(): Promise<void> {
     }
 
     const requestToProcess = persistedRequests[0];
+    processingRequest = requestToProcess;
 
     // Set the current request to a promise awaiting its processing so that getCurrentRequest can be used to take some action after the current request has processed.
     currentRequest = Request.processWithMiddleware(requestToProcess, true)
@@ -131,6 +133,7 @@ function flush() {
                 isSequentialQueueRunning = false;
                 resolveIsReadyPromise?.();
                 currentRequest = null;
+                processingRequest = null;
                 flushOnyxUpdatesQueue();
             });
         },
@@ -161,7 +164,7 @@ NetworkStore.onReconnection(flush);
 
 function push(request: OnyxRequest) {
     // Add request to Persisted Requests so that it can be retried if it fails
-    PersistedRequests.save(request);
+    PersistedRequests.save(request, processingRequest);
 
     // If we are offline we don't need to trigger the queue to empty as it will happen when we come back online
     if (NetworkStore.isOffline()) {
diff --git a/src/libs/actions/PersistedRequests.ts b/src/libs/actions/PersistedRequests.ts
index 3d37a22749..5ab5c1ae68 100644
--- a/src/libs/actions/PersistedRequests.ts
+++ b/src/libs/actions/PersistedRequests.ts
@@ -1,5 +1,6 @@
 import isEqual from 'lodash/isEqual';
 import Onyx from 'react-native-onyx';
+import HttpUtils from '@libs/HttpUtils';
 import ONYXKEYS from '@src/ONYXKEYS';
 import type {Request} from '@src/types/onyx';
 
@@ -17,7 +18,7 @@ function clear() {
     return Onyx.set(ONYXKEYS.PERSISTED_REQUESTS, []);
 }
 
-function save(requestToPersist: Request) {
+function save(requestToPersist: Request, processingRequest: Request | null) {
     const requests = [...persistedRequests];
     let hasConflict = false;
 
@@ -29,7 +30,10 @@ function save(requestToPersist: Request) {
         hasConflict = conflictingRequests.length > 0;
 
         conflictingRequests.forEach((conflictingRequest) => {
-            // delete the conflicting request
+            // delete and cancel the conflicting request
+            if (conflictingRequest === processingRequest) {
+                HttpUtils.cancelPendingRequests();
+            }
             const index = requests.findIndex((req) => req === conflictingRequest);
             if (index !== -1) {
                 requests.splice(index, 1);
@@ -41,7 +45,7 @@ function save(requestToPersist: Request) {
     }
 
     if (!shouldExcludeNewRequestOnConflict || !hasConflict) {
-        requests.push(requestToPersist)
+        requests.push(requestToPersist);
     }
 
     // Don't try to serialize conflict resolution functions

```

</details>
@roryabraham @shubham1206agra What do you guys think?

**Comment by bernhardoj** [comment]
@shubham1206agra I'm thinking of that alternative too, but that means both AddComment and DeleteComment requests of the same action will be processed which would probably cause https://github.com/Expensify/App/issues/36351 again.

**Comment by bernhardoj** [comment]
_I tested it and the request is cancelled btw_

**Comment by roryabraham** [comment]
> I am unsure if request cancellation will even work once the request is sent

I don't think it will work, because even though the HTTPS response is cancelled/not applied, Pusher events may still be triggered by the API call. So I suspect what might happen with your diff above @bernhardoj is:

1. User A sends a comment then quickly deletes it
1. `AddComment` has been sent but no response has been received, and we cancel the request along with the `DeleteComment` request.
1. The server processes the `AddComment` request and sends it to User B, causing User B to see the comment that should have been deleted

File: src/libs/actions/PersistedRequests.ts [hunk 1]
```diff
@@ -18,16 +18,15 @@ function clear() {
 }
 
 function save(requestToPersist: Request) {
-    const requests = [...persistedRequests, requestToPersist];
+    const requests = [...persistedRequests];
+    let hasConflict = false;
 
     // identify and handle any existing requests that conflict with the new one
-    const {getConflictingRequests, handleConflictingRequest, shouldIncludeCurrentRequest} = requestToPersist;
+    const {getConflictingRequests, handleConflictingRequest, shouldExcludeNewRequestOnConflict} = requestToPersist;
```
Comment (roryabraham) [inline src/libs/actions/PersistedRequests.ts:25]: I see you flipped this such that the default is that the new request _is_ included by default. We should change that back such that the new request is unaffected unless you also pass `shouldIncludeNewRequestOnConflict`

**Review by roryabraham — CHANGES_REQUESTED** [review]
@bernhardoj to fix the latest issue you reported, I think we should:

- move this logic up out of `libs/actions/PersistedRequests` and into `src/libs/Network/SequentialQueue`
- exclude the `currentRequest` from being considered in any conflicts. Once it's sent we should not try to take it back.

Thanks, and let us know if you have further questions

**Comment by bernhardoj** [comment]
@roryabraham do you mean to move the whole conflicting request logic to here?
https://github.com/Expensify/App/blob/aed0bf0e015d745f84f2961955f715bde73ec073/src/libs/Network/SequentialQueue.ts#L162-L165

**Comment by roryabraham** [comment]
@bernhardoj yep

**Comment by roryabraham** [comment]
We are treating this as a fire now, so I'm going to continue this PR over here: https://github.com/Expensify/App/pull/40059

**Pull request closed without merging**